util/EIP-712/src/encode.rs

```diff
@@ -0,0 +1,356 @@
+			}
+
+			keccak(items).0.to_vec()
+		}
+
+		Type::Custom(ref ident) if message_types.get(&*ident).is_some() => {
+			let type_hash = (&type_hash(ident, &message_types)?).to_vec();
+			let mut tokens = encode(&[EthAbiToken::FixedBytes(type_hash)]);
+
+			for field in message_types.get(ident).expect("Already checked in match guard; qed") {
+				let value = &value[&field.name];
+				let type_ = parser.parse_type(&*field.type_)?;
+				let mut encoded = encode_data(parser, &type_, &message_types, &value, Some(&*field.name))?;
+				tokens.append(&mut encoded);
+			}
+
+			keccak(tokens).0.to_vec()
+		}
+
+		Type::Bytes(size) if *size == 32 => {
+			let string = value.as_str().ok_or_else(|| serde_error("string", field_name))?;
+			if string.len() <= 2 {
+				return Err(ErrorKind::HexParseError(
+					format!("Expected a 0x-prefixed string of even length, found {} length string", string.len()))
+				)?
+			}
+			let string = string.get(2..).expect("line 188 checks for length; qed");
+			let bytes = H256::from_str(string).map_err(|err| ErrorKind::HexParseError(format!("{}", err)))?;
+			let hash = (&keccak(&bytes)).to_vec();
+			encode(&[EthAbiToken::FixedBytes(hash)])
+		}
+
+		Type::Bytes(size) if *size < 32 => {
+			let string = value.as_str().ok_or_else(|| serde_error("string", field_name))?;
+			if string.len() <= 2 {
+				return Err(ErrorKind::HexParseError(
+					format!("Expected a 0x-prefixed string of even length, found {} length string", string.len()))
+				)?
+			}
+			let string = string.get(2..).expect("line 200 checks for length; qed");
+			let bytes = H256::from_str(string).map_err(|err| ErrorKind::HexParseError(format!("{}", err)))?;
+			encode(&[EthAbiToken::FixedBytes(bytes.to_vec())])
+		}
+
+		Type::String => {
+			let value = value.as_str().ok_or_else(|| serde_error("string", field_name))?;
+			let hash = (&keccak(value)).to_vec();
+			encode(&[EthAbiToken::FixedBytes(hash)])
+		}
+
+		Type::Bool => encode(&[EthAbiToken::Bool(value.as_bool().ok_or_else(|| serde_error("bool", field_name))?)]),
+
+		Type::Address => {
+			let addr = value.as_str().ok_or_else(|| serde_error("string", field_name))?;
+			if addr.len() != 42 {
+				return Err(ErrorKind::InvalidAddressLength(addr.len()))?;
+			}
+			// we've checked the address length, this is safe
+			let addr = addr.get(2..).unwrap();
+			let address = EthAddress::from_str(addr).map_err(|err| ErrorKind::HexParseError(format!("{}", err)))?;
+			encode(&[EthAbiToken::Address(address)])
+		}
+
+		Type::Uint | Type::Int => {
+			// try to deserialize as a number first, then a string
+			let uint = match (value.as_u64(), value.as_str()) {
+				(Some(number), _) => U256::from(number),
+				(_, Some(string)) => {
+					if string.len() <= 2 {
+						return Err(ErrorKind::HexParseError(
+							format!("Expected a 0x-prefixed string of even length, found {} length string", string.len()))
+						)?
+					}
+					let string = string.get(2..).expect("line 200 checks for length");
```

Comment:
I wouldn't recommend refering checks by line, because of maintenance burden imposed on the future editors. Chances are that nobody actually will notice breaking this invariant by adding some lines before it.

In fact, this already is outdated since line 200 is occupied by this `/// encodes and hashes the given EIP712 struct`.